### PR TITLE
update pose_dock, add rigorous work and dock_and_equilibrate

### PR DESCRIPTION
**Pull request by schmolly**

### CODE REVIEW

File: docking/dock_and_equilibrate.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+        x0 = np.concatenate([solvated_host_coords, orig_guest_coords])
+        v0 = np.zeros_like(x0)
+        print(
+            f"SYSTEM", f"guest_name: {guest_name}", f"num_atoms: {len(x0)}",
+        )
+
+        for atom_num in constant_atoms:
+            combined_masses[atom_num - 1] += 50000
+
+        seed = 2020
```
Comment (proteneer) [inline docking/dock_and_equilibrate.py:163]: Maybe we should make this 2021 :)

File: docking/dock_and_equilibrate.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+                        outdir,
+                        guest_name,
+                        str(step).zfill(len(str(insertion_steps))),
+                        f"ins",
+                    )
+                if too_much_force(ctxt, combined_bps, host_box, u_impls, lamb):
+                    calc_work = False
+                    break
+
+        if (
```
Comment (proteneer) [inline docking/dock_and_equilibrate.py:204]: Note that this condition is only true for ABFE - if we do RBFE this condition would no longer hold.

File: docking/dock_and_equilibrate.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+    print(
+        f"{stage}\t"
+        f"guest_name: {guest_name}\t"
+        f"step: {str(step).zfill(len(str(n_steps)))}\t"
+        f"lambda: {lamb:.2f}\t"
+        f"energy: {energy:.2f}"
+    )
+
+
+def too_much_force(ctxt, bps, box, u_impls, lamb):
```
Comment (proteneer) [inline docking/dock_and_equilibrate.py:259]: This function should be called sparingly. It is *hideously* expensive due to all the memcpys. Even once every 100 steps might be too often.

File: docking/pose_dock.py [hunk 4]
```diff
@@ -66,11 +70,40 @@ def pose_dock(
...
+    host_nb_bp = None
+    for bp in host_potentials:
+        if isinstance(bp, potentials.Nonbonded):
+            # (ytz): hack to ensure we only have one nonbonded term
+            assert host_nb_bp is None
+            host_nb_bp = bp
+        else:
+            final_potentials.append(bp)
+
+    padding = 0.1
```
Comment (proteneer) [inline docking/pose_dock.py:98]: note to self and @maxentile  we should really fix this later on. This padding was done to address the particles that are too close to the boundary.

File: docking/pose_dock.py [hunk 5]
```diff
@@ -281,18 +309,16 @@ def pose_dock(
         "--max_lambda",
         type=float,
         default=1.1,
```
Comment (proteneer) [inline docking/pose_dock.py:311]: Lambda is now strictly between 0 and 1.

File: docking/dock_and_equilibrate.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+    constant_atoms=[]
+):
+    """Solvates a host, inserts guest(s) into solvated host, equilibrates
+
+    Parameters
+    ----------
+
+    host_pdbfile: path to host pdb file to dock into
+    guests_sdfile: path to input sdf with guests to pose/dock
+    max_lambda: lambda value the guest should insert from or delete to
```
Comment (proteneer) [inline docking/dock_and_equilibrate.py:40]: Lambda is now strictly between 0 and 1.

File: tests/test_recipe.py [hunk 7]
```diff
@@ -22,7 +22,7 @@ def test_recipe_from_rdkit():
```
Comment (proteneer) [inline tests/test_recipe.py:22]: Once this is merged I will delete this file and the recipe class altogether.

File: docking/pose_dock.py (hunk 4 above)
Comment (maxentile) [reply docking/pose_dock.py:103]: Agreed!

**Review by maxentile — APPROVED** [review]
Great work! And thanks for walking through this on Thursday. Looks good to me.

**Review by proteneer — APPROVED** [review]

**Comment by proteneer** [comment]
LGTM. Feel free to merge.

**Pull request merged**